nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-nar-utils/src/main/java/org/apache/nifi/nar/StandardExtensionDiscoveringManager.java

```diff
@@ -145,74 +152,103 @@ public void discoverExtensions(final Set<Bundle> narBundles) {
+                        }
                     }
                 }
+            } catch (final IOException e) {
+                throw new RuntimeException("Failed to get resources of type " + serviceType + " from bundle " + bundle);
             }
-
-            classLoaderBundleLookup.put(bundle.getClassLoader(), bundle);
         }
+
+        classLoaderBundleLookup.put(bundle.getClassLoader(), bundle);
     }
 
-    protected void loadExtension(final Object extension, final Class<?> extensionType, final Bundle bundle) {
-        final boolean isControllerService = ControllerService.class.equals(extensionType);
-        final boolean isProcessor = Processor.class.equals(extensionType);
-        final boolean isReportingTask = ReportingTask.class.equals(extensionType);
+    private Set<String> getServiceFileImplementationClassNames(final URL serviceFileUrl) throws IOException {
+        final Set<String> implementationClassNames = new HashSet<>();
 
-        // create a cache of temp ConfigurableComponent instances, the initialize here has to happen before the checks below
-        if ((isControllerService || isProcessor || isReportingTask) && extension instanceof ConfigurableComponent) {
-            final ConfigurableComponent configurableComponent = (ConfigurableComponent) extension;
-            initializeTempComponent(configurableComponent);
+        try (final InputStream in = serviceFileUrl.openStream();
+             final Reader inputStreamReader = new InputStreamReader(in);
+             final BufferedReader reader = new BufferedReader(inputStreamReader)) {
 
-            final String cacheKey = getClassBundleKey(extension.getClass().getCanonicalName(), bundle.getBundleDetails().getCoordinate());
-            tempComponentLookup.put(cacheKey, configurableComponent);
-        }
+            String line;
+            while ((line = reader.readLine()) != null) {
+                // Remove anything after the #
+                final int index = line.indexOf("#");
+                if (index >= 0) {
+                    line = line.substring(0, index);
+                }
 
-        // only consider extensions discovered directly in this bundle
-        boolean registerExtension = bundle.getClassLoader().equals(extension.getClass().getClassLoader());
+                // Ignore empty line
+                line = line.trim();
+                if (line.isEmpty()) {
+                    continue;
+                }
 
-        if (registerExtension) {
-            final Class<?> extensionClass = extension.getClass();
-            if (isControllerService && !checkControllerServiceEligibility(extensionClass)) {
-                registerExtension = false;
-                logger.error(String.format(
-                    "Skipping Controller Service %s because it is bundled with its supporting APIs and requires instance class loading.", extensionClass.getName()));
+                implementationClassNames.add(line);
```

Comment:
The JDK `ServiceLoader` includes a check of the first character in a line that calls `Character.isJavaIdentifierStart()`, is it worth introducing such as check at this point?  The result of the check could throw an exception indicating an invalid line in the service file, as opposed to throwing an error somewhere higher in the chain when attempting to instantiate an class from an invalid string.

Reply:
in https://github.com/palindromicity/bundles, which is my breaking out of the nar system, I switched from service loader to https://github.com/atteo/classindex which is faster and have other benefits ( like marking services at build time so you don't have to manually maintain the service file ).  You may want to think about it

Reply:
@exceptionfactory we could add a check like that, but I don't know that it really buys us anything. It seems very arbitrary to me, to check one character on each line. Unless you have some deeper understanding that I am lacking, as to why that is an important check, I don't think it's worth muddying the codebase to try to discover an anomaly with the first character of each line, in particular, when any other character could be an issue as well

Reply:
@ottobackwards I'm not familiar with that project. Will have to look into it. Thanks.

Reply:
Thanks for the reply.  The Character is check appears to be just a sanity check in ServiceLoader.  Invalid class names will throw other errors, so it is really just a matter of where NiFi catches the problem.  Leaving out the check is fine if it is reasonable to wait until the class gets loaded to indicate what is most likely a typo.